images/hub/jupyterhub_config.py

```diff
@@ -120,87 +113,106 @@
+        ('oauth_callback_url', 'callback-url'),
+        ('hosted_domain', 'hosted-domain'),
+        ('login_service', 'login-service'),
+    ):
+        set_config_if_not_none(c.GoogleOAuthenticator, trait, 'auth.google.' + cfg_key)
     email_domain = get_config('auth.google.hosted-domain')
 elif auth_type == 'github':
-    c.JupyterHub.authenticator_class = 'oauthenticator.GitHubOAuthenticator'
-    c.GitHubOAuthenticator.oauth_callback_url = get_config('auth.github.callback-url')
-    c.GitHubOAuthenticator.client_id = get_config('auth.github.client-id')
-    c.GitHubOAuthenticator.client_secret = get_config('auth.github.client-secret')
-    org_whitelist = get_config('auth.github.org_whitelist', [])
-    if len(org_whitelist) != 0:
-        c.GitHubOAuthenticator.github_organization_whitelist = org_whitelist
+    c.JupyterHub.authenticator_class = 'oauthenticator.github.GitHubOAuthenticator'
+    for trait, cfg_key in (
+        ('client_id', 'client-id'),
+        ('client_secret', 'client-secret'),
+        ('oauth_callback_url', 'callback-url'),
+        ('github_organization_whitelist', 'org_whitelist'),
+    ):
+        set_config_if_not_none(c.GitHubOAuthenticator, trait, 'auth.github.' + cfg_key)
 elif auth_type == 'cilogon':
     c.JupyterHub.authenticator_class = 'oauthenticator.CILogonOAuthenticator'
-    c.CILogonOAuthenticator.oauth_callback_url = get_config('auth.cilogon.callback-url')
-    c.CILogonOAuthenticator.client_id = get_config('auth.cilogon.client-id')
-    c.CILogonOAuthenticator.client_secret = get_config('auth.cilogon.client-secret')
+    for trait, cfg_key in (
+        ('client_id', 'client-id'),
+        ('client_secret', 'client-secret'),
+        ('oauth_callback_url', 'callback-url'),
+    ):
+        set_config_if_not_none(c.CILogonOAuthenticator, trait, 'auth.cilogon.' + cfg_key)
 elif auth_type == 'gitlab':
     c.JupyterHub.authenticator_class = 'oauthenticator.gitlab.GitLabOAuthenticator'
-    c.GitLabOAuthenticator.oauth_callback_url = get_config('auth.gitlab.callback-url')
-    c.GitLabOAuthenticator.client_id = get_config('auth.gitlab.client-id')
-    c.GitLabOAuthenticator.client_secret = get_config('auth.gitlab.client-secret')
+    for trait, cfg_key in (
+        ('client_id', 'client-id'),
+        ('client_secret', 'client-secret'),
+        ('oauth_callback_url', 'callback-url'),
+    ):
+        set_config_if_not_none(c.GitLabOAuthenticator, trait, 'auth.gitlab.' + cfg_key)
 elif auth_type == 'mediawiki':
     c.JupyterHub.authenticator_class = 'oauthenticator.mediawiki.MWOAuthenticator'
-    c.MWOAuthenticator.client_id = get_config('auth.mediawiki.client-id')
-    c.MWOAuthenticator.client_secret = get_config('auth.mediawiki.client-secret')
-    c.MWOAuthenticator.index_url = get_config('auth.mediawiki.index-url')
+    for trait, cfg_key in (
+        ('client_id', 'client-id'),
+        ('client_secret', 'client-secret'),
+        ('oauth_callback_url', 'callback-url'),
```

Comment:
Is this a bugfix in this authenticator configuration? There are no deleted entry about oauth_callback_url.

Perhaps we need to fix the configmap as well if so

Reply:
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/f5dd68cf7e6ddec4579b1c8fa87b56ddb56a59ce/jupyterhub/templates/hub/configmap.yaml#L72-L76

Yepp, so either we should not set `oauth_callback_url` here, or update the configmap to accept such entry as well.

Reply:
Fortunately, this is why this PR exists! Failing to set oauth callback URL will have no negative consequences, because it's only set if the value is provided. This is exactly the kind of bug we are now resilient to after this PR.

But yes, it is a bug that the oauth callback URL isn't exposed for this authenticator.